src/megaapi_impl.cpp

```diff
@@ -21503,82 +21400,49 @@ void MegaApiImpl::sendPendingRequests()
             const char *name = request->getName();
 
             auto nextSyncTag = client->nextSyncTag();
-            MegaSyncPrivate *sync = new MegaSyncPrivate(localPath, name, node->nodehandle, nextSyncTag);
+            unique_ptr<MegaSyncPrivate> sync(new MegaSyncPrivate(localPath, name, node->nodehandle, nextSyncTag));
             sync->setRegExp(request->getRegExp());
 
-            SyncError syncError = NO_SYNC_ERROR;
-
             std::unique_ptr<char []> remotePath{getNodePathByNodeHandle(request->getNodeHandle())};
             if (!remotePath)
             {
                 e = API_ENOENT;
-                syncError = REMOTE_NODE_NOT_FOUND;
-                request->setNumDetails(syncError);
+                request->setNumDetails(REMOTE_NODE_NOT_FOUND);
                 break;
             }
 
             SyncConfig syncConfig{nextSyncTag, localPath, name, request->getNodeHandle(), remotePath.get(),
                                   0, regExpToVector(request->getRegExp())};
 
-            e = client->addsync(syncConfig, DEBRISFOLDER, NULL, syncError, true, sync);
-            request->setNumDetails(syncError);
-            if (!e || !isSyncErrorPermanent(syncError))
+            UnifiedSync* unifiedSync = nullptr;
+            e = client->addsync(syncConfig, DEBRISFOLDER, NULL, true, unifiedSync);
+            request->setNumDetails(syncConfig.getError());
+            if (unifiedSync)
             {
-                auto newstate = isSyncErrorPermanent(syncError) ? SYNC_FAILED : SYNC_DISABLED;
-                sync->setState(newstate);
-                if (!e)
+                if (!e && unifiedSync && unifiedSync->mSync)
                 {
-                    Sync *s = client->syncs.back();
-                    fsfp_t fsfp = s->fsfp;
-                    sync->setState(s->state); //override state with the actual one from the sync
+                    fsfp_t fsfp = unifiedSync->mSync->fsfp;
                     sync->setLocalFingerprint(fsfp);
                     request->setNumber(fsfp);
                 }
-                sync->setError(syncError);
                 sync->setMegaFolderYielding(remotePath.release());
                 request->setTransferTag(nextSyncTag);
-                syncMap[nextSyncTag] = sync;
 
-                fireOnSyncAdded(sync, e ? MegaSync::NEW_TEMP_DISABLED : MegaSync::NEW);
+                fireOnSyncAdded(sync.get(), e ? MegaSync::NEW_TEMP_DISABLED : MegaSync::NEW);
```

Comment:
Note, this is happening now after fireOnSyncStateChanged: 
```1  mega::MegaApiImpl::fireOnSyncStateChanged megaapi_impl.cpp 16638 0x55dee8db1715 
2  mega::MegaApiImpl::syncupdate_stateconfig megaapi_impl.cpp 13093 0x55dee8d9c7d3 
3  mega::UnifiedSync::changedConfigState     sync.cpp         2146  0x55dee8cc10c4 
4  mega::UnifiedSync::enableSync             sync.cpp         1996  0x55dee8cc035b 
5  mega::MegaClient::addsync                 megaclient.cpp   13110 0x55dee8c4f452 
...
```
which unfortunately breaks the regular flow the app shall expect :-(

Reply:
Hi, from the state of the code I had thought that is how it was supposed to work - so that the app can track the state of all syncs just from callbacks for one consistent method.  Or should it be that we don't notify of a change if the app asked for the change itself?   I started making code alterations for that second way (by all relevant functions taking a `bool notifyApp` parameter) but then thought, better to clarify first :-)  thanks

Reply:
I'm not sure we are in the same page here:. 
Changes should always be notified and onSyncStateChange should be the one consistent method, but the app is expecting first to receive:
fireOnSyncAdded (to start tracking the sync), and then the corresponding state fireOnSyncStateChanged if changes occur.  With the revamping, the state changes happen before the first onSyncAdded arrives. Am I making sense?

Reply:
Ah I see, thanks for explaining the concern, that makes sense.   I've made a small adjustment (a bool notifyApp parameter to addsync() ) that addresses the issue, and prevents fireOnSyncStateChanged notifications before fireOnSyncAdded.  Apps like megacli using the SDK directly may prefer to be pass true to be notified immediately, but intermediate layer apps pass false. thanks

Reply:
Could we better add a new member to the `UnifiedSync::isInitialized` or alike, that helps the `fireOnSyncStateChanged()` to decide whether the notification should reach the app or not? I think it would be cleaner than the flag `notifyApp` in these 3 methods passing around.

Reply:
Nevermind, I've seen that Pablo already replaced the flag by a lambda completion function :)